File: 0101-0200/0142.Linked_List_Cycle_II.java
//0ms



/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) break;
        }
        if (fast == null || fast.next == null) return null;
        while (head != slow) {
            head = head.next;
            slow = slow.next;
        }
        return head;
    }
}




//0ms




public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while(fast!=null && fast.next!=null){
            fast=fast.next.next;
            slow=slow.next;
            if(slow==fast){
                return startingPoint(head,fast,slow);
            }
        }
         return null;
        
    }
    public ListNode startingPoint(ListNode head, ListNode fast, ListNode slow){
        slow=head;
            while(fast!=slow){
                fast=fast.next;
                slow=slow.next;

            }
            return slow;
    }
}





//0ms






/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        int len = length(head);
        if(len==-1)
        {
            return null;
        }
        ListNode f = head;
        ListNode s = head;
        while(len>0)
        {
            f=f.next;
            len--;
        }
        while(f!=s)
        {
            f=f.next;
            s=s.next;
        }
        return s;
    }
    int length(ListNode head)
    {
        ListNode fast = head;
        ListNode slow = head;        
        while(fast!=null && fast.next!=null)
        {
            fast = fast.next.next;
            slow = slow.next;
            if(fast==slow)
            {
                int length = 0;
                ListNode temp = slow;
                do
                {
                    length++;
                    temp = temp.next;
                }while(temp!=slow);
                return length;
            }
        }
        return -1;
    }
}

File: 0101-0200/0142.Linked_List_Cycle_II.cpp
//7ms




/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) break;
        }
        if (!(fast && fast->next)) return NULL;
        while (head != slow) {
            head = head->next;
            slow = slow->next;
        }
        return head;
    }
};




//3ms





/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
          if(head==NULL || head->next==NULL) return NULL;
        ListNode* slow=head;
        ListNode* fast=head;
        ListNode* n=nullptr;
        int fl=0;

        while(fast!=nullptr&&fast->next!=nullptr)
        {
            slow=slow->next;
            fast=fast->next->next;
            if(slow==fast)
            {
                fl=1;
                break;
            }
        }
        if(fl==0)
        return n;
        slow=head;
        while(slow!=fast)
        {
            slow=slow->next;
            fast=fast->next;

        }
        return slow;
        
    }
};





//10ms





/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow=head;ListNode* fast=head;
        if(head==NULL || head->next==NULL)
        return NULL;
        if(head->next->next==head){
            return head;
        }
        while(fast!=NULL && fast->next!=NULL ){
            slow=slow->next;
            fast=fast->next->next;
            if(slow==fast){
                slow=head;
                while(slow!=fast){
                    slow=slow->next;
                fast=fast->next;

                }
                return slow;
                

            }
            
        }
        return NULL;
        
        
    }
};





//5ms





/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;

        while(fast && fast -> next){
            slow = slow -> next;
            fast = fast -> next -> next;

            if(slow == fast){
                slow = head;

                while(slow != fast){
                    slow = slow -> next;
                    fast = fast -> next;
                }
                return slow;
            }
        }
        return NULL;
    }
};
